compiler/src/iree/compiler/Preprocessing/Common/PadToIntrinsics.cpp

```diff
@@ -172,45 +290,271 @@ static void padConvOp(RewriterBase &rewriter, linalg::LinalgOp linalgOp,
   rewriter.replaceOp(linalgOp, extracted);
 }
 
-struct PadToIntrinsicsPass final
-    : impl::PadToIntrinsicsBase<PadToIntrinsicsPass> {
-  void runOnOperation() override {
-    MLIRContext *context = &getContext();
-    RewritePatternSet patterns(context);
-    Operation *rootOp = getOperation();
-    ArrayAttr mmaKinds = nullptr;
-    for (IREE::HAL::ExecutableTargetAttr targetAttr :
-         IREE::HAL::DeviceTargetAttr::lookupExecutableTargets(rootOp)) {
-      FailureOr<ArrayAttr> candidateMmaKinds =
-          getSupportedMmaTypes(targetAttr.getConfiguration());
-      if (succeeded(candidateMmaKinds)) {
-        mmaKinds = *candidateMmaKinds;
-        break;
+static void padContractionLikeOp(RewriterBase &rewriter,
+                                 linalg::LinalgOp linalgOp) {
+  FailureOr<mlir::linalg::ContractionDimensions> contractionDims =
+      mlir::linalg::inferContractionDims(linalgOp);
+
+  if (failed(contractionDims)) {
+    return;
+  }
+
+  if (contractionDims->k.size() < 1 || contractionDims->m.size() < 1 ||
+      contractionDims->n.size() < 1) {
+    return;
+  }
+
+  // Early exit if cannot find intrinsics or if multiple executable targets.
+  SmallVector<GPUMatmulShapeType> intrinsics = getIntrinsics(linalgOp);
+  if (intrinsics.empty())
+    return;
+
+  Location loc = linalgOp.getLoc();
+
+  // Naive handling by only looking into most inner dimensions.
+  int64_t mDim = contractionDims->m.back();
+  int64_t nDim = contractionDims->n.back();
+  int64_t kDim = contractionDims->k.back();
+
+  // If none of the shape is dynamic, we'd fallback to using pad to intrinsics.
+  SmallVector<int64_t, 4> bounds = linalgOp.getStaticLoopRanges();
+  int64_t mSize = bounds[mDim];
+  int64_t nSize = bounds[nDim];
+  int64_t kSize = bounds[kDim];
+
+  // Bail out on matvec-like cases.
+  if (mSize == 1 || nSize == 1) {
+    return;
+  }
+
+  // Util to get operand and dim;
+  auto getSrcOperandAndDim =
+      [&](int64_t targetDim) -> std::optional<std::pair<Value, int64_t>> {
+    for (auto operand : linalgOp.getDpsInputOperands()) {
+      auto operandMap = linalgOp.getMatchingIndexingMap(operand);
+      std::optional<unsigned> maybeDim = operandMap.getResultPosition(
+          getAffineDimExpr(targetDim, operandMap.getContext()));
+      if (maybeDim)
+        return std::make_pair(operand->get(), maybeDim.value());
```

Comment:
nit: you can construct pair directly, `make_pair` is mostly a c++98 thing now :P

```suggestion
        return std::pair{operand->get(), maybeDim.value()};
```

Reply:
done!